### ISSUE TITLE
[ENH]: generalize artists opting in-to auto limits

### ISSUE THREAD

**Issue by tacaswell** [matplotlib/matplotlib#25139]
### Problem

This is triggered by #25127 most recently, but this is a discussion we have had a couple of times.

The core of the problem is that users have different expectations about what things will automatically adjust the axes limits (or not) in autolim mode.  Currently `Lines`, `Patches` and `AxesImage` obligatorily participate https://github.com/matplotlib/matplotlib/blob/a046ee3febac399869c3eb6b15b8e669e37058ee/lib/matplotlib/axes/_base.py#L2459-L2483 and collections may optionally participate, but only when they are added https://github.com/matplotlib/matplotlib/blob/a046ee3febac399869c3eb6b15b8e669e37058ee/lib/matplotlib/axes/_base.py#L2260-L2273 

### Proposed solution

The proposed solution is to:

 - move the `_update_line_limits` and friends to the respective Artists
 - add an analogous method to the base `Artist` (probably defaulting to failure)
 - add a "I would like to particpate in autolimiting!" flag to base `Artist`
 - in `relim` look at the flag and call the newly generalized method above on any artists that opt-in

This will involved a little bit of public API (how to set the state to opt-in) and a bunch of private API (what should the signature of the method be).

**Comment by jklymak** [comment]
A major difficulty is that many of our Artists may be anchored in data co-ordinates, but their bounding boxes are not in data-coordinates.

**Comment by tacaswell** [comment]
True, but now we can stuff that complexity back to the `Artist` classes where we have the best chance of knowing how to actually resolve that problem.

**Comment by anntzer** [comment]
See #15595, which I had tried to push for during a sprint.

**Comment by tacaswell** [comment]
oh, nice that even has some good discussion over the tricky parts of the API (that I participated in and have no memory of 😞 )!

**Comment by jayyxiao** [comment]
I am trying to use #15595 as a starting point for this issue. It implemented an autolim flag in base Artist, `_in_autoscale`. It was not merged and I believe it is inactive. Can I include the existing code in my PR?

Also, for the `update_*_limits` in base Artist, how should it default to failure? Should `update_*_limits` immediately raise an Error, so it will fail if it is not overriden by a child class? What type of Error would it be?
